### ISSUE TITLE
Wrong autograd when weight vector is multiplied by a large CSR matrix

### ISSUE THREAD

**Issue by kai-lan** [pytorch/pytorch#99848]
### 🐛 Describe the bug

When a weight vector is multiplied by a relatively large CSR matrix, the gradient computed seems wrong.

The example matrix `A` is tri-diagonal with `2`s on main diagonal and `-1`s on sub-diagonals. The weight vector `x` is randomly generated. 

I then tried to verify the gradient by finite difference validation. I plotted $\|\frac{Loss(x +\epsilon) - Loss(x-\epsilon)}{2\epsilon} - \frac{dLoss}{dx} \|/\| \frac{dLoss}{dx}\|$ VS $\epsilon$ on log-log scale. If the gradient is correct, we should expect a slope of 2. But this examples fails when `n=50000` and with CSR tensor. If we convert CSR to COO, it would give the correct result.

Expected: 
![finite_difference_correct](https://user-images.githubusercontent.com/59628235/233877381-afe2c669-c4de-4b09-8608-e4ea2795cc2a.png)
Result: 

![finite_difference_wrong](https://user-images.githubusercontent.com/59628235/233877397-bd37b0e0-b42c-4cb4-99c5-b0033f54136e.png)

```
import scipy.sparse as sparse
import numpy as np
import torch
import torch.nn as nn
import matplotlib.pyplot as plt
torch.manual_seed(4)

def lap1d(n, dtype=np.float64):
    v  = np.ones(n, dtype=dtype)
    L1 = sparse.spdiags([-v,2*v,-v],[-1,0,1],n,n)
    return L1

n = 50000 # 40000 would be fine
A_sp = lap1d(n).tocsr()

A = torch.sparse_csr_tensor(A_sp.indptr, A_sp.indices, A_sp.data, A_sp.shape, dtype=torch.float64)
# A = A.to_sparse_coo() # If converted to COO it would be fine.
x = nn.Parameter(torch.rand(n, dtype=torch.float64))

def loss(x, A):
    x = nn.functional.silu(x)
    return (A @ x).norm()

numerical = []
perturb = np.logspace(-8, 0, 50)
with torch.no_grad():
    for eps in perturb:
        L_p = loss(x + eps, A)
        L_m = loss(x - eps, A)
        numerical.append((L_p - L_m) / (2*eps))
numerical = np.array(numerical)

L = loss(x, A)
L.backward()
dLdx = x.grad.sum().item()
err = np.abs(numerical - dLdx) / np.abs(dLdx)

plt.loglog(perturb, err)
plt.xlabel('epsilon')
plt.ylabel('Relative error')
plt.grid()
plt.savefig("finite_difference.png")
```

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0.post200
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.2
Libc version: glibc-2.31

Python version: 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 15:55:03)  [GCC 10.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 525.105.17
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          32
On-line CPU(s) list:             0-31
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           33
Model name:                      AMD Ryzen 9 5950X 16-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         2200.000
CPU max MHz:                     3400.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        6787.37
Virtualization:                  AMD-V
L1d cache:                       512 KiB
L1i cache:                       512 KiB
L2 cache:                        8 MiB
L3 cache:                        64 MiB
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[pip3] torch-scatter==2.1.0
[pip3] torch-sparse==0.6.16
[pip3] torchaudio==0.13.1
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.14.1
[conda] cudatoolkit               11.8.0              h37601d7_11    conda-forge
[conda] magma                     2.6.2                hc72dce7_0    conda-forge
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] numpy                     1.23.5           py39h3d75532_0    conda-forge
[conda] pytorch                   1.13.0          cuda112py39hb0b7ed5_200    conda-forge
[conda] pytorch-cuda              11.7                 h67b0de4_1    pytorch
[conda] pytorch-scatter           2.1.0           py39_torch_1.13.0_cu117    pyg
[conda] pytorch-sparse            0.6.16          py39_torch_1.13.0_cu117    pyg
[conda] torch                     1.13.1                   pypi_0    pypi
[conda] torchaudio                0.13.1                   pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.14.1                   pypi_0    pypi
```

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

**Comment by nikitaved** [comment]
@kai-lan , does it reproduce with the latest PyTorch? We had quite a stream of fixes for sparse-related mms...

**Comment by alexsamardzic** [comment]
With version 2.0.0+cu118, indeed the result turns out as expected.

**Comment by kai-lan** [comment]
Thanks for your reply! I just updated to latest version of pytorch, and this problem indeed has been resolved.
